File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    let itr;
    let itr_end = matrix.length;
    

    let fnd = 0;
    for(itr=0; itr<itr_end; itr++) {
        if (!fnd) {
            let start = 0;
            let end = matrix[0].length;

            while (start <= end) {
                let mid = start + Math.floor((end-start)/2);

                if (matrix[itr][mid] == target) {
                    fnd = 1;
                    break;
                } else if (matrix[itr][mid]<target) start = mid+1;
                else end = mid-1;
            }
        }
    }

    return fnd === 1;
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n, m = len(matrix), len(matrix[0])
        left, right = 0, n*m-1
        while left <= right:
            mid = left + (right-left)//2
            if matrix[mid//m][mid%m] == target: return True
            elif matrix[mid//m][mid%m] < target: left = mid+1
            else: right = mid-1
        
        return False

